# A/B Test Analysis
**Question:** Does the new checkout raise the payment rate among users who start checkout?
**Test:** Two-proportion z-test, 95% confidence.

In [ ]:
import pandas as pd
from sqlalchemy import create_engine
from statsmodels.stats.proportion import proportions_ztest, confint_proportions_2indep

engine = create_engine("postgresql+psycopg2://postgres:YOUR_PASSWORD@localhost:5432/funnel_doctor")

df = pd.read_sql("SELECT * FROM experiment_results WHERE started_checkout = 1", engine)

summary = df.groupby("variant")["paid"].agg(["sum", "count"])
summary["pay_rate"] = summary["sum"] / summary["count"]
print(summary)

new_paid, new_n = int(summary.loc["new", "sum"]), int(summary.loc["new", "count"])
old_paid, old_n = int(summary.loc["old", "sum"]), int(summary.loc["old", "count"])

z, p_value = proportions_ztest([new_paid, old_paid], [new_n, old_n])
low, high = confint_proportions_2indep(new_paid, new_n, old_paid, old_n)

print("Lift (new - old):", round(summary.loc["new", "pay_rate"] - summary.loc["old", "pay_rate"], 4))
print("p-value:", p_value)
print("95% confidence interval for lift:", round(low, 4), "to", round(high, 4))

          sum  count  pay_rate
variant                       
new      1742   3111  0.559949
old      1586   3066  0.517286
Lift (new - old): 0.0427
p-value: 0.0007709299244884571
95% confidence interval for lift: 0.0178 to 0.0675


**Result:** New 56.0% vs old 51.7% (+4.3 points, 95% CI 1.8 to 6.8, p < 0.001). The difference is statistically significant.

In [7]:
for device, g in df.groupby("device"):
    s = g.groupby("variant")["paid"].agg(["sum", "count"])
    if len(s) < 2:
        continue
    new_paid, new_n = int(s.loc["new", "sum"]), int(s.loc["new", "count"])
    old_paid, old_n = int(s.loc["old", "sum"]), int(s.loc["old", "count"])

    z, p = proportions_ztest([new_paid, old_paid], [new_n, old_n])
    low, high = confint_proportions_2indep(new_paid, new_n, old_paid, old_n)

    print(f"{device}: old {old_paid/old_n:.3f} ({old_n} users), "
          f"new {new_paid/new_n:.3f} ({new_n} users), "
          f"lift {new_paid/new_n - old_paid/old_n:.3f}, "
          f"p-value {p:.4f}, CI {low:.3f} to {high:.3f}")

Unknown: old 0.504 (115 users), new 0.583 (103 users), lift 0.078, p-value 0.2475, CI -0.054 to 0.206
desktop: old 0.650 (1097 users), new 0.676 (1131 users), lift 0.026, p-value 0.1868, CI -0.013 to 0.066
mobile: old 0.435 (1714 users), new 0.486 (1731 users), lift 0.051, p-value 0.0026, CI 0.018 to 0.084
tablet: old 0.500 (140 users), new 0.521 (146 users), lift 0.021, p-value 0.7282, CI -0.094 to 0.135


**By device:** Mobile improved significantly (+5.1 points, p = 0.003). Desktop, tablet and unknown showed small gains that were not significant.
**Recommendation:** Roll out the new checkout, mainly for mobile.
**Limitations:** Synthetic data with a built-in effect; smaller groups within each device; novelty effects could not be measured.